pkg/codegen/schema/schema.go

```diff
@@ -0,0 +1,898 @@
+	// Properties is a map from property name to PropertySpec that describes the type's properties.
+	Properties map[string]PropertySpec `json:"properties,omitempty"`
+	// Type must be "object".
+	Type string `json:"type,omitempty"`
+	// Requires is a list of the names of the type's required properties. These properties must be set for inputs and
+	// will always be set for outputs.
+	Required []string `json:"required,omitempty"`
+	// Language specifies additional language-specific data about the type.
+	Language map[string]json.RawMessage `json:"language,omitempty"`
+}
+
+// AliasSpec is the serializable form of an alias description.
+type AliasSpec struct {
+	// Name is the name portion of the alias, if any.
+	Name *string `json:"name,omitempty"`
+	// Project is the project portion of the alias, if any.
+	Project *string `json:"project,omitempty"`
+	// Type is the type portion of the alias, if any.
+	Type *string `json:"type,omitempt"`
+}
+
+// ResourceSpec is the serializable form of a resource description.
+type ResourceSpec struct {
+	ObjectTypeSpec
+
+	// InputProperties is a map from property name to PropertySpec that describes the resource's input properties.
+	InputProperties map[string]PropertySpec `json:"inputProperties,omitempty"`
+	// Defauls is a map from property name to DefaultSpec that describes the default values for this resource's input
+	// properties.
+	Defaults map[string]DefaultSpec `json:"defaults,omitempty"`
+	// RequiredInputs is a list of the names of the resource's required input properties.
+	RequiredInputs []string `json:"requiredInputs,omitempty"`
+	// StateInputs is an optional ObjectTypeSpec that describes additional inputs that mau be necessary to get an
+	// existing resource. If this is unset, only an ID is necessary.
+	StateInputs *ObjectTypeSpec `json:"stateInputs,omitempty"`
+	// Aliases is the list of aliases for the resource.
+	Aliases []AliasSpec `json:"aliases,omitempty"`
+	// DeprecationMessage indicates whether or not the resource is deprecated.
+	DeprecationMessage string `json:"deprecationMessage,omitempty"`
+	// Language specifies additional language-specific data about the resource.
+	Language map[string]json.RawMessage `json:"language,omitempty"`
+}
+
+// FunctionSpec is the serializable form of a function description.
+type FunctionSpec struct {
+	// Description is the description of the function, if any.
+	Description string `json:"description,omitempty"`
+	// Inputs is the bag of input values for the function, if any.
+	Inputs *ObjectTypeSpec `json:"inputs,omitempty"`
+	// Outputs is the bag of output values for the function, if any.
+	Outputs *ObjectTypeSpec `json:"outputs,omitempty"`
+	// DeprecationMessage indicates whether or not the function is deprecated.
+	DeprecationMessage string `json:"deprecationMessage,omitempty"`
+	// Language specifies additional language-specific data about the function.
+	Language map[string]json.RawMessage `json:"language,omitempty"`
```

Comment:
i skimmed these.  are there meaninful differences in teh hierarchy for types and the hierarchy for typespecs?  if so, would love to understand those.

Reply:
Not sure what you're asking here.